Refactor configuration parsers to use Lark

micaeljtoliveira opened the pull request.

edoyango commented on src/access/parsers/config.py:
```diff
@@ -0,0 +1,420 @@
+    def __setitem__(self, key: str, value: Any) -> None:
+        """Override method to set item from dict.
+
+        This method takes care of updading the parse tree, so that when writing it back into text it will use the new
```
```suggestion
        This method takes care of updating the parse tree, so that when writing it back into text it will use the new
```

edoyango commented on src/access/parsers/config.py:
```diff
@@ -0,0 +1,420 @@
+
+        # Currently we only support replacing existing values, not adding new ones
+        if key not in self:
+            raise (KeyError("Key already exist: ", key))
```
```suggestion
            raise (KeyError(f"Key doesn't exist: {key}"))
```
changes the error message from `('Key already exist: ', 'potato')`
to `KeyError: "Key doesn't exist: potato"`
If adding keys will be supported later on, maybe `NotImplemented` error could be raised instead?

micaeljtoliveira replied:
Not entirely sure, but I don't think adding keys will be supported. It involves quite a lot of code (creating an entire tree branch for the new item) for limited use-cases.

edoyango commented on src/access/parsers/config.py:
```diff
@@ -0,0 +1,420 @@
+class ConfigParser(ABC):
+    """Lark-based configuration parser base class.
+
+    The parsers build by extending this class are meant for files that share a common structure. In this case, the
```
```suggestion
    The parsers built by extending this class are meant for files that share a common structure. In this case, the
```

edoyango commented on src/access/parsers/config.py:
```diff
@@ -0,0 +1,420 @@
+        return str(value)
+
+
+_value_sanity_check = {
```
Since this is being used in if statements and such, maybe it was be easier to read if it was named something like `_is_same_type`?

micaeljtoliveira replied:
Good idea. Initially it include some other sanity checks, but as the code evolved it ended up by only containing type checks, so makes sense to rename.

edoyango commented on src/access/parsers/config.py:
```diff
@@ -0,0 +1,420 @@
+            raise (KeyError("Key already exist: ", key))
+
+        if isinstance(value, dict):
+            raise (SyntaxError("Trying to assing a new value to an entire block"))
```
```suggestion
            raise (SyntaxError("Trying to assign a new value to an entire block"))
```

edoyango commented on src/access/parsers/fortran_nml.py:
```diff
@@ -0,0 +1,63 @@
+class FortranNMLParser(ConfigParser):
+    """Fortran Namelist parser.
+
+    Note: Currently array qualifiers, substrings and derived types are not implementd in the grammar.
```
```suggestion
    Note: Currently array qualifiers, substrings and derived types are not implemented in the grammar.
```

minghangli-uni reviewed (changes requested):
Thanks @micaeljtoliveira this is fantastic! I’m not very familiar with Lark, and it seems there’s quite a bit to take in. I’ll need to spend more time reading through it properly.

I did a quick test using only the Fortran nml parser in the `experiment-generator`. I suggested a small change for missing a `raise` before `TypeError` and a change so that namelist groups can also terminate with `&end` as well as `/`.

I also noticed that the current parser struggles with multi-line assignments, for example:
```
    fields_from_atm   = 'swfld_i', 'lwfld_i', 'rain_i', 'snow_i', 'press_i',
                        'runof_i', 'tair_i', 'qair_i', 'uwnd_i', 'vwnd_i',
                        'licalvf_i'
```

One more thing: is support for adding a new key already a to-do plan for the parser?

minghangli-uni commented:
I haven’t tested the other modules yet. Would you mind if I run some tests and get back to you tomorrow?

Also, do you think it might make sense to rename `payu_config_yaml.py` to something more general, like `yaml_config.py`? That way it’s less confusing for users. I’m not sure about all use cases, but in the OM model the `diag_table` is generated from [`diagnostic_profiles/source_yaml_files/diag_table_standard_source.yaml`](https://github.com/ACCESS-NRI/access-om3-configs/blob/dev-MC_25km_jra_ryf/diagnostic_profiles/source_yaml_files/diag_table_standard_source.yaml), which I believe your YAML parser should also be able to handle.

micaeljtoliveira commented:
> I haven’t tested the other modules yet. Would you mind if I run some tests and get back to you tomorrow?

Sure, no worries. Those kind of tests are very useful to find corner cases not implemented in the grammars.

> Also, do you think it might make sense to rename `payu_config_yaml.py` to something more general, like `yaml_config.py`? That way it’s less confusing for users. I’m not sure about all use cases, but in the OM model the `diag_table` is generated from [`diagnostic_profiles/source_yaml_files/diag_table_standard_source.yaml`](https://github.com/ACCESS-NRI/access-om3-configs/blob/dev-MC_25km_jra_ryf/diagnostic_profiles/source_yaml_files/diag_table_standard_source.yaml), which I believe your YAML parser should also be able to handle.

Yes, I agree. These parsers shouldn't check if the contents of the files are valid for each file format, only if they are well-formed (that is, they follow the grammar rules). That means the YAML parser should be completely generic.

If later on we want to implement validation for specific file formats like payu configs, this should be implemented in a different module.

micaeljtoliveira commented:
@minghangli-uni I've now renamed the YAML related files from "payu_config_yaml" to "yaml_config".

micaeljtoliveira commented:
@edoyango Thanks for the feedback! Yesterday I pushed some changes that should fix the multiline lists. Could you check it works now? I'll add the trailing comma ASAP.

micaeljtoliveira commented:
> I'll add the trailing comma ASAP.

Done.

micaeljtoliveira commented:
> Yep that works! Could you also allow for whitespaces before the group and "/"? `atmosphere/namelists` has a space before everything for some reason.

Done.

edoyango commented:
> Can you share the code you are using to test this?

Sure!
```python
from access.parsers.fortran_nml import FortranNMLParser 
parser = FortranNMLParser()
with open("access-esm1.6-configs/atmosphere/namelists", "r") as f:
    txt = "".join(f.readlines())
c = parser.parse(txt)
print(c)
```
If it helps, I'm using lark 1.2.2 and python 3.10

micaeljtoliveira commented:
Well, took me a while to find it, but lark is 100% correct: that comment is in the middle of the file!!!

```bash
$ grep -C5 "End of file" namelists 
 0,0,0,0,0,0,0,0,0,
 SSTAnom='N',
 AOBINC= 0 ,
 AOBGRP= 0 ,
 /
  ## End of file ##
 &NLSTCALL
 EXPT_ID='aiih',
 JOB_ID='c',
 EXPT_ID_IN='    ',
 JOB_ID_IN=' ',
```

I don't know if this is legal Fortran or not, but the comment is clearly wrong :laughing: So I would suggest we remove it instead of "fixing" the parser.

micaeljtoliveira commented:
> Oh good find! Turns out there's multiple "END OF FILE" comments haha. It looks like the hashes have been there since the config's inception so it's in all the branches. I'll post an issue, and we'll see what they say.

Thanks! I can change the parser to accept anything outside the blocks, but I would first like to know if that's legal or not. I'll try to investigate this. I just hope this is clearly stated in the standard and not left to the compilers...

> I was also wrong about it being the last problem too :(
> 
> ```
>  Num_ALBCs=,
> ```
> 
> trailing comma in empty assignment produces an error.

Oh, this one is going to be annoying. Null assignments are currently not supported by the parsers. I was hoping this wouldn't show up, but I should have known better.

minghangli-uni commented:
I did some simple tests, unfortunately only the yaml and fortran nml parsers work for now. 

fortran nml 

```python
parser = fortran_nml.FortranNMLParser()
text = Path("/g/data/tm70/ml0072/COMMON/git_repos/access-experiment-generator/new_parsers/edge_cases/1deg_jra55_ryf/ice/cice_in.nml").read_text()
nml_parser = parser.parse(text)
del nml_parser['setup_nml']['dbug']
nml_parser
```

The deletion worked correctly. However, the output shows that all group and key names are converted to uppercase,

```
{'SETUP_NML': {'DAYS_PER_YEAR': 365,
  'DIAG_FILE': 'ice_diag.d',
  'DIAG_TYPE': 'file',
  'DIAGFREQ': 960,
...
```
Since fortran nml keys are case-insensitive, I'd expect this wont cause anything wrong. 

For nuopc.runconfig, 
```python
parser = nuopc_config.NUOPCParser()
text = Path("/g/data/tm70/ml0072/COMMON/git_repos/access-om3-configs/nuopc.runconfig").read_text()
nuopc_parser = parser.parse(text)
nuopc_parser
```

```
UnexpectedCharacters: No terminal matches '[/](https://are.nci.org.au/)' in the current parser context, at line 6 col 21

     outPathRoot = [./](https://are.nci.org.au/node/gadi-cpu-bdw-0002.gadi.nci.org.au/23972/lab/tree/g/data/tm70/ml0072/COMMON/git_repos/sandbox/access-parsers_understanding/)
                    ^
Expected one of: 
	* __ANON_0
	* COLON
	* NEWLINE
	* config__WS_INLINE
```

For MOM_input,

```python
parser = mom6_input.MOM6InputParser()
text = Path("/g/data/tm70/ml0072/COMMON/git_repos/access-om3-configs/MOM_input").read_text()
# print(text)
mom_parser = parser.parse(text)
```
comes with an error

```
UnexpectedCharacters: No terminal matches ' ' in the current parser context, at line 495 col 1

                                ! If tru
^
Expected one of: 
	* PERCENT
	* config__CNAME
```

micaeljtoliveira commented:
@minghangli-uni Thanks for the testing and feedback.

> I did some simple tests, unfortunately only the yaml and fortran nml parsers work for now.
> 
> fortran nml
> 
> ```python
> parser = fortran_nml.FortranNMLParser()
> text = Path("/g/data/tm70/ml0072/COMMON/git_repos/access-experiment-generator/new_parsers/edge_cases/1deg_jra55_ryf/ice/cice_in.nml").read_text()
> nml_parser = parser.parse(text)
> del nml_parser['setup_nml']['dbug']
> nml_parser
> ```
> 
> The deletion worked correctly. However, the output shows that all group and key names are converted to uppercase,
>
> ```
> {'SETUP_NML': {'DAYS_PER_YEAR': 365,
>   'DIAG_FILE': 'ice_diag.d',
>   'DIAG_TYPE': 'file',
>   'DIAGFREQ': 960,
> ...
> ```
> 
> Since fortran nml keys are case-insensitive, I'd expect this wont cause anything wrong.

No, it won't cause any issues. Internally the keys are stored in uppercase, but the dict access is case-insensitive. So doing:
```python
nml_parser["Setup"]["diag_File"]
```
will work correctly.

 
> For nuopc.runconfig,
> 
> ```python
> parser = nuopc_config.NUOPCParser()
> text = Path("/g/data/tm70/ml0072/COMMON/git_repos/access-om3-configs/nuopc.runconfig").read_text()
> nuopc_parser = parser.parse(text)
> nuopc_parser
> ```
> 
> ```
> UnexpectedCharacters: No terminal matches '[/](https://are.nci.org.au/)' in the current parser context, at line 6 col 21
> 
>      outPathRoot = [./](https://are.nci.org.au/node/gadi-cpu-bdw-0002.gadi.nci.org.au/23972/lab/tree/g/data/tm70/ml0072/COMMON/git_repos/sandbox/access-parsers_understanding/)
>                     ^
> Expected one of: 
> 	* __ANON_0
> 	* COLON
> 	* NEWLINE
> 	* config__WS_INLINE
> ```

I hadn't encountered a url in those files before, so didn't implement it. Should be straightforward to add.

> For MOM_input,
> 
> ```python
> parser = mom6_input.MOM6InputParser()
> text = Path("/g/data/tm70/ml0072/COMMON/git_repos/access-om3-configs/MOM_input").read_text()
> # print(text)
> mom_parser = parser.parse(text)
> ```
> 
> comes with an error
> 
> ```
> UnexpectedCharacters: No terminal matches ' ' in the current parser context, at line 495 col 1
> 
>                                 ! If tru
> ^
> Expected one of: 
> 	* PERCENT
> 	* config__CNAME
> ```

This one should be trivial to fix.

minghangli-uni commented:
> I hadn't encountered a url in those files before, so didn't implement it. Should be straightforward to add.

Just to clarify, the error is coming from the line `outPathRoot = ./`. I’m unsure why it’s picking up my jupyter notebook url there…

```
DRIVER_attributes::
     Verbosity = off
     cime_model = cesm
     drv_restart_pointer = rpointer.cpl
     logFilePostFix = .log
     outPathRoot = ./
```

micaeljtoliveira commented:
> > I hadn't encountered a url in those files before, so didn't implement it. Should be straightforward to add.
> 
> Just to clarify, the error is coming from the line `outPathRoot = ./`. I’m unsure why it’s picking up my jupyter notebook url there…
> 
> ```
> DRIVER_attributes::
>      Verbosity = off
>      cime_model = cesm
>      drv_restart_pointer = rpointer.cpl
>      logFilePostFix = .log
>      outPathRoot = ./
> ```

Ah, right, that makes more sense. This is also trivial to fix then.

micaeljtoliveira commented:
@edoyango I think I fixed all the issues with the Fortran namelists and I'm now able to parse `atmosphere/namelists` without any errors. Would you mind having another look whenever you've got some time?

@minghangli-uni I also fixed most of the issues you reported. Currently only key addition is missing I think.

micaeljtoliveira commented:
@edoyango I partially fixed the last problem you reported. Now the namelist won't get ignored, but only that last instance will be taken into account. I'll add the proper handling of namelists appearing multiple times in a given file in a different PR if that's okay. It's not too hard to fix, but requires a few more changes.
